eole/decoders/transformer.py

```diff
@@ -254,16 +255,26 @@ def update_dropout(self, dropout, attention_dropout):
         for layer in self.transformer_layers:
             layer.update_dropout(dropout, attention_dropout)
 
-    def _causal_attn_mask(self, tgt_pad_mask):
-        tgt_len = tgt_pad_mask.size(-1)
+    def _causal_attn_mask(self, tgt_pad_mask, pfxlen=None):
+        batch_size, _, tgt_len = tgt_pad_mask.size()
+        device = tgt_pad_mask.device
         # Add triangular future_mask and pad_mask, result mask in (B, T, T).
         future_mask = torch.tril(
             torch.ones((tgt_len, tgt_len), device=tgt_pad_mask.device, dtype=torch.bool),
             diagonal=0,
-        )
+        )  # Shape: (tgt_len, tgt_len)
         if self.sliding_window > 0:
             future_mask = future_mask.triu_(-self.sliding_window)
-        attn_mask = ~tgt_pad_mask & future_mask.unsqueeze(0)
+        if self.prefix_LM and pfxlen is not None:
+            idx = torch.arange(tgt_len, device=device)
+            block = (idx.view(1, tgt_len, 1) < pfxlen.view(batch_size, 1, 1)) & (
+                idx.view(1, 1, tgt_len) < pfxlen.view(batch_size, 1, 1)
+            )
+            future_mask = torch.where(block, torch.ones_like(future_mask), future_mask)  # (B, T, T)
```

Comment:
This looks fairly similar to the vision logic (full attention on image, causal on the rest), might be good to factorize in some utility function at some point for clarity.